docs/middleware.md

```diff
@@ -0,0 +1,76 @@
+# Using Next 12 middleware on Netlify
+
+Next 12 introduces a new feature called [middleware](https://nextjs.org/docs/middleware), which are functions that are
+run before a request has finished processing, and can be used to modify the request or replace the response. For
+example, it can change headers, rewrite the request path, or return a different response entirely.
+
+Next Middleware can run either in an edge function or at the origin. On Netlify, middleware runs at the origin as part
+of the normal Next.js server.
+
+## How to deploy Next 12 middleware
+
+Next 12 Middleware works out of the box with Netlify, and most functions will work unchanged. See
+[the middleware docs](https://nextjs.org/docs/middleware) for details of how to create them. There are however a few
+workarounds that are currently required for some features during the beta period:
+
+### `geo`
+
+When running at the origin, Next does not populate the `request.geo` object. Fortunately there is a one line fix to get
+the visitor's country:
+
+```typescript
+export async function middleware(req: NextRequest) {
+  // Add this line
+  req.geo.country = req.headers.get('x-country')
+
+  // The rest of your middleware goes here
+}
+```
+
+### `ip`
+
+Next.js also does not populate the `req.ip` value when running at the origin. There is another one line fix for this:
+
+```typescript
+export async function middleware(req: NextRequest) {
+  // Add this line
+  req.ip = req.headers.get('x-nf-client-connection-ip')
+
+  // The rest of your middleware goes here
+}
+```
+
+## Caveats
+
+Because the middleware runs at the origin, it is run _after_ Netlify rewrites and redirects. If a static file is served
+then it is never run, as it is only run when a page is served by Next.js. This means that it should not be used with the
+`EXPERIMENTAL_MOVE_STATIC_FILES` option, as this causes statically-generated pages to be served by the Netlify CDN
+before any middleware can be run.
```

Comment:
Making some guesses and some edits to make this a tiny bit clearer:

```suggestion
Because the middleware runs at the origin, it is run _after_ Netlify rewrites and redirects. If a static file is served by the Netlify CDN 
then the middleware is never run, as middleware only runs when a page is served by Next.js. This means that middleware should not be used with the
`EXPERIMENTAL_MOVE_STATIC_FILES` option, as this causes statically-generated pages to be served by the Netlify CDN
before any middleware can be run.
```